File: src/Web/Dart/example3.dart
// Abstract Product
abstract class Report {
  void generate();
}

// Concrete Product
class PDFReport implements Report {
  @override
  void generate() {
    print('Generating PDF report');
  }
}

class HTMLReport implements Report {
  @override
  void generate() {
    print('Generating HTML report');
  }
}

// Abstract Factory
abstract class ReportFactory {
  Report createReport();
}

// Concrete Factory
class PDFReportFactory implements ReportFactory {
  @override
  Report createReport() {
    return PDFReport();
  }
}

class HTMLReportFactory implements ReportFactory {
  @override
  Report createReport() {
    return HTMLReport();
  }
}

void useFactory(ReportFactory factory) {
  var report = factory.createReport();
  report.generate();
}

// Usage
void main() {
  useFactory(PDFReportFactory());
  useFactory(HTMLReportFactory());
}

File: src/Web/TypeScriptTS/example3.ts
// Abstract Product
interface Report {
    generate(): void;
}

// Concrete Product
class PDFReport implements Report {
    generate(): void {
        console.log("Generating PDF report");
    }
}

class HTMLReport implements Report {
    generate(): void {
        console.log("Generating HTML report");
    }
}

// Abstract Factory
interface ReportFactory {
    createReport(): Report;
}

// Concrete Factory
class PDFReportFactory implements ReportFactory {
    createReport(): Report {
        return new PDFReport();
    }
}

class HTMLReportFactory implements ReportFactory {
    createReport(): Report {
        return new HTMLReport();
    }
}

function useFactory(factory: ReportFactory): void {
    const report = factory.createReport();
    report.generate();
}

// Usage
useFactory(new PDFReportFactory());
useFactory(new HTMLReportFactory());
